dtls.c

```diff
@@ -2643,16 +2713,45 @@ dtls_send_client_hello(dtls_context_t *ctx, dtls_peer_t *peer,
 
     dtls_int_to_uint8(p, TLS_EXT_EC_POINT_FORMATS_UNCOMPRESSED);
     p += sizeof(uint8);
+
+    /* signature algorithms extension */
+    dtls_int_to_uint16(p, TLS_EXT_SIG_HASH_ALGO);
+    p += sizeof(uint16);
+
+    /* length of this extension type */
+    dtls_int_to_uint16(p, 4);
+    p += sizeof(uint16);
+
+    /* supported_signature_algorithms */
+    dtls_int_to_uint16(p, 2);
+    p += sizeof(uint16);
+
+    /* sha256 */
+    dtls_int_to_uint8(p, TLS_EXT_SIG_HASH_ALGO_SHA256);
+    p += sizeof(uint8);
+
+    /* ecdsa */
+    dtls_int_to_uint8(p, TLS_EXT_SIG_HASH_ALGO_ECDSA);
+    p += sizeof(uint8);
+
+    /* extended master secret */
+    dtls_int_to_uint16(p, TLS_EXT_EXTENDED_MASTER_SECRET);
+    p += sizeof(uint16);
+
+    /* length of this extension type */
+    dtls_int_to_uint16(p, 0);
+    p += sizeof(uint16);
+    handshake->extended_master_secret = 1;
+
   }
 
   assert((buf <= p) && ((unsigned int)(p - buf) <= sizeof(buf)));
 
-  if (cookie_length != 0)
-    clear_hs_hash(peer);
+  clear_hs_hash(peer);
 
   return dtls_send_handshake_msg_hash(ctx, peer, &peer->session,
 				      DTLS_HT_CLIENT_HELLO,
-				      buf, p - buf, cookie_length != 0);
+				      buf, p - buf, 1);
```

Comment:
I am having a hard time wrapping my head around this change. It looks as if you decided that independent of the cookie's presence a new hash calculation is started at this point

Reply:
Not all servers return a hello_verify (gnutls did not until I fixed that in libcoap).  As there was no second client_hello, client_hello never never made it into the hash.  This is the reason why I now always reset the hash in client_hello so that we only get one client_hello in the hash.

Reply:
So far, that sounds reasonable. I was just wondering if this might affect other invocations of `dtls_send_client_hello()` where cookie_length is explicitly set to 0, e.g., in `dtls_renegotiate()`?

Reply:
Interesting question.  During this code development, I did a quick test of #23 to check I was not making that worse - and found that it worked.  However, when I rechecked at the end of the coding/testing, #23 was not working again, but I could not get to to work again - a change like this may have made a difference here - I will try to test that further to understand what is really happening there in #23

Reply:
This does not break dtls_renegotiation().  #23 is now fixed in PR #41